packages/medusa/src/services/csv-parser.ts

```diff
@@ -0,0 +1,99 @@
+import { AwilixContainer } from "awilix"
+import Papa, { ParseConfig } from "papaparse"
+import { AbstractParser } from "../interfaces/abstract-parser"
+import { CsvSchema } from "../interfaces/csv-parser"
+
+const DEFAULT_PARSE_OPTIONS = {
+  dynamicTyping: true,
+  header: true,
+}
+
+class CsvParser<
+  TSchema extends CsvSchema = CsvSchema,
+  TParserResult = unknown,
+  TOutputResult = unknown
+> extends AbstractParser<TSchema, TParserResult, ParseConfig, TOutputResult> {
+  protected readonly $$delimiter: string = ";"
+
+  constructor(
+    protected readonly container: AwilixContainer,
+    schema: TSchema,
+    delimiter?: string
+  ) {
+    super(schema)
+    if (delimiter) {
+      this.$$delimiter = delimiter
+    }
+  }
+
+  public async parse(
+    readableStream: NodeJS.ReadableStream,
+    options: ParseConfig = DEFAULT_PARSE_OPTIONS
+  ): Promise<TParserResult[]> {
+    const csvStream = Papa.parse(Papa.NODE_STREAM_INPUT, options)
+
+    const parsedContent: TParserResult[] = []
+    readableStream.pipe(csvStream)
+    for await (const chunk of csvStream) {
+      parsedContent.push(chunk)
+    }
+
+    return parsedContent
+  }
+
+  async buildData(data: TParserResult[]): Promise<TOutputResult[]> {
+    const validatedData = [] as TOutputResult[]
+    for (let i = 0; i < data.length; i++) {
+      const builtLine = await this._buildLine(data[i], i + 1)
+      validatedData.push(builtLine)
+    }
+    return validatedData
+  }
+
+  private async _buildLine(
+    line: TParserResult,
+    lineNumber: number
+  ): Promise<TOutputResult> {
+    const outputTuple = {} as TOutputResult
+    const columnMap = this._buildColumnMap(this.$$schema.columns)
+
+    const tupleKeys = Object.keys(line)
+
+    for (const tupleKey of tupleKeys) {
```

Comment:
We should iterate over keys from the schema (`columnMap`) and check if defined columns are provided instead of `tupleKeys` from `line`. The current implementation doesn't cover the case where we have a (required) column defined in our schema and that column doesn't exist in provided data.

Reply:
to test your scheam and implementation, you could do a fake file with those columns https://github.com/medusajs/medusa/pull/1602/files#diff-d08d472b8f2619af22cc2dfda528c529463c1063ac42bf87aa65eff046918656R92 and test it on that and on a copy of that with less or more column with the same schema

Reply:
Thanks for pointing this out @fPolic, it's a very good point! I gave this a thought and I think that there are two options to solve this, especially if we want to include @adrien2p 's suggestion:
1. **Iterating over keys from the schema**: This solves the problem you mentioned in a nice way, but becomes a bit trickier when introducing this notion of "many entries/columns in a line can correspond to one column defined in the schema", for example: 
```javascript
{
  columns: [
    ...,
    {
      name: "variants",
      match: /.*Variant Price.*/i,
      transformer: (value) => Math.round(Number(value) * 100),
    },
  ],
}
```
It gets trickier here because now, when you reach this column, you have to find all properties in the line object which correspond to this column, then apply validation + transformation on each. I think we can make it work no problem, but I feel like option 2 might be a bit simpler/more intuitive

2. **Iterating over the tuple/line keys**: Similarly to what we've had so far, we'll add an extra step for resolving the column from the tupleKey _i.e: by knowing this is "variant price dkk", we should find some column that matches this name_. Then, the algorithm still behaves the same way it used to before this change. 
Now to address the valid point you raised, we can keep track of the columns we've seen in a "visited" map, then once we finish processing the line, we simply do 1 pass over the schema columns and make sure we've visited all of them when we were processing the line, if any column is missing, we throw an error. 

Apologies if this is not all super clear, this comment is supposed to complement the latest commit I've pushed. I have went with option 2, and this comment summarizes the thought process behind it as well as justifies why i've went with option 2. Hope it makes sense 😄

Reply:
I understand you point and maybe that for the first iteration what we have it quite nice with the option 2. 

I suggested that because, when we export the data, some columns are dynamic and it can be 1-N of then. Then if you want to re import it to update the data for exemple, you have to update the schema, but its fine i guess

Reply:
Yes with the march and the look throught the tuple columns and the visited idea everything is align 😅 nice work man one shot 🚀